### PR TITLE
Policy Feature implementation

### LINKED ISSUE #102: Create AddPolicy Command

### LINKED ISSUE #104: Add Policy field into Person

### PR DESCRIPTION
**Pull request by ijavierja**
close #102 #103 #104 #105 #106 #107 #108

### CODE REVIEW

**Review by Joven-Heng — APPROVED** [review]
LGTM

**Review by WeiJie96 — CHANGES_REQUESTED** [review]
Just some small changes

**Review by WeiJie96 — APPROVED** [review]
LGTM

File: src/main/java/seedu/address/storage/JsonAdaptedPerson.java [hunk 1]
```diff
@@ -103,6 +106,12 @@ public JsonAdaptedPerson(Person source) {
         priority = source.getPriority().value;
         assert priority != null : "Value stored in priority should not be null";
 
+        if (source.getPolicy() != null) {
+            policy = new JsonAdaptedPolicy(source.getPolicy());
+            assert policy != null : "Value stored in address should not be null";
```
Comment (WeiJie96) [inline src/main/java/seedu/address/storage/JsonAdaptedPerson.java:111]: Perhaps the assertion can print "Policy should not be null"
Perhaps we should check policy.getPolicyName() != null and policy.getPolicyDescription() != null as well
and print "Policy name should not be null" and "Policy description should not be null"

File: src/test/java/seedu/address/logic/commands/AddCommandTest.java [hunk 2]
```diff
@@ -65,26 +70,35 @@ public void execute_duplicatePerson_throwsCommandException() {
...
+    }
+
     @Test
     public void equals() {
         Person alice = new PersonBuilder().withName("Alice").build();
         Person bob = new PersonBuilder().withName("Bob").build();
-        AddCommand addAliceCommand = new AddCommand(alice);
-        AddCommand addBobCommand = new AddCommand(bob);
+        AddCommand addAliceCommand = new AddCommand(alice, null);
+        AddCommand addBobCommand = new AddCommand(bob, null);
```
Comment (WeiJie96) [inline src/test/java/seedu/address/logic/commands/AddCommandTest.java:96]: Perhaps we should test for adding policy as well

File: src/test/java/seedu/address/model/ModelManagerTest.java [hunk 3]
```diff
@@ -149,25 +182,24 @@ public void equals() {
...
         differentUserPrefs.setClientListFilePath(Paths.get("differentFilePath"));
-        assertFalse(modelManager.equals(new ModelManager(clientList, differentUserPrefs)));
+        assertFalse(modelManager.equals(new ModelManager(clientList, differentUserPrefs, policyList)));
 
         // different isArchiveMode -> returns false
-        ModelManager modelManagerCopyInArchiveMode = new ModelManager(clientList, userPrefs);
+        ModelManager modelManagerCopyInArchiveMode = new ModelManager(clientList, userPrefs, policyList);
         modelManagerCopyInArchiveMode.setIsArchiveMode(true);
         assertFalse(modelManager.equals(modelManagerCopyInArchiveMode));
-
```
Comment (WeiJie96) [inline src/test/java/seedu/address/model/ModelManagerTest.java:171]: Perhaps we should test for different policyList -> returns false

File: src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java [hunk 4]
```diff
@@ -140,27 +229,46 @@ public void toModelType_invalidAddress_throwsIllegalValueException() {
...
+                VALID_PHONE,
+                VALID_EMAIL,
+                VALID_ADDRESS,
+                VALID_CLIENTSOURCES,
+                VALID_NOTE,
+                VALID_IS_ARCHIVE,
+                INVALID_PRIORITY,
+                new JsonAdaptedPolicy(VALID_POLICY_NAME, VALID_POLICY_DESCRIPTION));
         assertThrows(IllegalValueException.class, person::toModelType);
     }
```
Comment (WeiJie96) [inline src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java:273]: Perhaps we should have invalidPolicy test, with both invalid name and description

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 5]
```diff
@@ -97,42 +97,42 @@ public void parse_optionalFieldsMissing_success() {
         Person expectedPersonWithoutPhone = new PersonBuilder(AMY).withoutPhone().build();
         assertParseSuccess(parser, NAME_DESC_AMY + EMAIL_DESC_AMY + ADDRESS_DESC_AMY
                         + CLIENTSOURCE_DESC_FRIEND + NOTE_DESC_CAT + PRIORITY_DESC_AMY,
-                new AddCommand(expectedPersonWithoutPhone));
+                new AddCommand(expectedPersonWithoutPhone, null));
```
Comment (WeiJie96) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:100]: Perhaps we should have Policy here, as we are testing one field at a time. Same for rest of optional fields missing

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 6]
```diff
@@ -97,42 +97,42 @@ public void parse_optionalFieldsMissing_success() {
...
                         + ADDRESS_DESC_AMY + CLIENTSOURCE_DESC_FRIEND + NOTE_DESC_CAT,
-                new AddCommand(expectedPersonWithoutPriority));
+                new AddCommand(expectedPersonWithoutPriority, null));
 
         // missing priority prefix
         assertParseSuccess(parser, NAME_DESC_AMY + PHONE_DESC_AMY + EMAIL_DESC_AMY
                 + ADDRESS_DESC_AMY + CLIENTSOURCE_DESC_FRIEND + NOTE_DESC_CAT,
-                new AddCommand(expectedPersonWithoutPriority));
+                new AddCommand(expectedPersonWithoutPriority, null));
     }
```
Comment (WeiJie96) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:136]: Perhaps we should test for missing Policy name prefix

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 7]
```diff
@@ -59,35 +59,35 @@ public void parse_allFieldsPresent_success() {
         // whitespace only preamble
         assertParseSuccess(parser, PREAMBLE_WHITESPACE + NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
                 + ADDRESS_DESC_BOB + CLIENTSOURCE_DESC_FRIEND + NOTE_DESC_DOG
-                + PRIORITY_DESC_BOB, new AddCommand(expectedPerson));
+                + PRIORITY_DESC_BOB, new AddCommand(expectedPerson, null));
```
Comment (WeiJie96) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:63]: Perhaps we should have policy here

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 8]
```diff
@@ -59,35 +59,35 @@ public void parse_allFieldsPresent_success() {
...
 
         // multiple clientsources - all accepted
         Person expectedPersonMultipleTags = new PersonBuilder(BOB).withClientSources(VALID_CLIENTSOURCE_FRIEND,
                 VALID_CLIENTSOURCE_HUSBAND)
                 .withNote(VALID_NOTE_DOG).build();
         assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ADDRESS_DESC_BOB
                         + CLIENTSOURCE_DESC_FRIEND + CLIENTSOURCE_DESC_HUSBAND + NOTE_DESC_DOG + PRIORITY_DESC_BOB,
-                new AddCommand(expectedPersonMultipleTags));
+                new AddCommand(expectedPersonMultipleTags, null));
     }
```
Comment (WeiJie96) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:91]: Perhaps we should test for multiple policy names - last policy name accepted

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 9]
```diff
@@ -59,35 +59,35 @@ public void parse_allFieldsPresent_success() {
...
         assertParseSuccess(parser, PREAMBLE_WHITESPACE + NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
                 + ADDRESS_DESC_BOB + CLIENTSOURCE_DESC_FRIEND + NOTE_DESC_DOG
-                + PRIORITY_DESC_BOB, new AddCommand(expectedPerson));
+                + PRIORITY_DESC_BOB, new AddCommand(expectedPerson, null));
 
         // multiple names - last name accepted
         assertParseSuccess(parser, NAME_DESC_AMY + NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
                 + ADDRESS_DESC_BOB + CLIENTSOURCE_DESC_FRIEND + NOTE_DESC_DOG
-                + PRIORITY_DESC_BOB, new AddCommand(expectedPerson));
+                + PRIORITY_DESC_BOB, new AddCommand(expectedPerson, null));
```
Comment (WeiJie96) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:67]: Perhaps we should have policy here, same for rest of multiple fields

File: src/test/java/seedu/address/logic/parser/AddPolicyCommandParserTest.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+                PREAMBLE_WHITESPACE + POLICY_NAME_DESC_AMY + POLICY_DESCRIPTION_DESC_AMY,
+                new AddPolicyCommand(expectedPolicy));
+    }
+
+    @Test
+    public void parse_missingPolicyFields_throwsParseException() {
+        assertThrows(ParseException.class, () -> parser.parse(PREAMBLE_WHITESPACE));
+        assertThrows(ParseException.class, () -> parser.parse(PREAMBLE_WHITESPACE + POLICY_NAME_DESC_AMY));
+        assertThrows(ParseException.class, () -> parser.parse(PREAMBLE_WHITESPACE + POLICY_DESCRIPTION_DESC_AMY));
+    }
```
Comment (WeiJie96) [inline src/test/java/seedu/address/logic/parser/AddPolicyCommandParserTest.java:33]: Perhaps we can refer to add command parser for some tests
- invalid value: failure
- all fields present success => multiple policy names / policy description - last policy name / policy description accepted

File: src/test/java/seedu/address/logic/commands/AddPolicyCommandTest.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+        Policy policy = LIFE_TIME;
+        Model model = new ModelManager();
+        Model expectedModel = new ModelManager();
+        expectedModel.addPolicy(LIFE_TIME);
+        String expectedMessage = String.format(AddPolicyCommand.MESSAGE_SUCCESS, policy);
+        assertCommandSuccess(new AddPolicyCommand(policy), model, expectedMessage, expectedModel);
+    }
+
+    @Test
+    public void execute_policyIsDuplicatePolicy() {
```
Comment (WeiJie96) [inline src/test/java/seedu/address/logic/commands/AddPolicyCommandTest.java:35]: Perhaps the method name should have throwsCommandException behind

File: src/test/java/seedu/address/logic/commands/AddPolicyCommandTest.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+        Model model = new ModelManager(new ClientList(), new UserPrefs(), getTypicalPolicyList());
+        assertCommandFailure(new AddPolicyCommand(LIFE_TIME), model, AddPolicyCommand.MESSAGE_DUPLICATE_POLICY);
+    }
```
Comment (WeiJie96) [inline src/test/java/seedu/address/logic/commands/AddPolicyCommandTest.java:38]: Perhaps we should have .equals() test

File: src/test/java/seedu/address/model/policy/PolicyListTest.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static seedu.address.testutil.Assert.assertThrows;
+import static seedu.address.testutil.TypicalPolicies.LIFE_TIME;
+import static seedu.address.testutil.TypicalPolicies.LIFE_TIME_NAME;
+import static seedu.address.testutil.TypicalPolicies.MISSING_FROM_LIST_POLICY;
+import static seedu.address.testutil.TypicalPolicies.MISSING_FROM_LIST_POLICY_NAME;
+import static seedu.address.testutil.TypicalPolicies.getTypicalPolicyList;
+
+import java.util.Hashtable;
+
```
Comment (WeiJie96) [inline src/test/java/seedu/address/model/policy/PolicyListTest.java:14]: Perhaps we should use HashMap here, as previously commented in PolicyList.java

File: src/test/java/seedu/address/model/policy/PolicyListTest.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+import org.junit.jupiter.api.Test;
+
+public class PolicyListTest {
+
+
+    @Test
+    public void constructor() {
+        PolicyList policyList = new PolicyList();
```
Comment (WeiJie96) [inline src/test/java/seedu/address/model/policy/PolicyListTest.java:22]: Perhaps we should have standardised names for the tests. Same for the rest of the tests in this file

File: src/test/java/seedu/address/model/policy/PolicyListTest.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+    public void getHashTableCopy() {
+        //check if its a copy
+        PolicyList policyList = getTypicalPolicyList();
+        policyList.getHashtableCopy().clear();
+        assertFalse(policyList.equals(new PolicyList()));
+
+        policyList.getHashtableCopy().put(MISSING_FROM_LIST_POLICY_NAME, MISSING_FROM_LIST_POLICY);
+        assertFalse(policyList.contains(MISSING_FROM_LIST_POLICY));
+    }
+
```
Comment (WeiJie96) [inline src/test/java/seedu/address/model/policy/PolicyListTest.java:74]: Perhaps we should test for 
- getPolicy(): E.g. null input, null output (key not present), key present
- equals()

File: src/test/java/seedu/address/model/policy/PolicyListTest.java [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    @Test
+    public void getHashTableCopy() {
+        //check if its a copy
+        PolicyList policyList = getTypicalPolicyList();
+        policyList.getHashtableCopy().clear();
+        assertFalse(policyList.equals(new PolicyList()));
+
+        policyList.getHashtableCopy().put(MISSING_FROM_LIST_POLICY_NAME, MISSING_FROM_LIST_POLICY);
+        assertFalse(policyList.contains(MISSING_FROM_LIST_POLICY));
```
Comment (WeiJie96) [inline src/test/java/seedu/address/model/policy/PolicyListTest.java:72]: Perhaps we should have a positive test

File: src/test/java/seedu/address/model/policy/PolicyListTest.java [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+        assertFalse(typicalPolicyList.contains(MISSING_FROM_LIST_POLICY_NAME));
+    }
+
+    @Test
+    public void add() {
+        PolicyList policyList = new PolicyList();
+        policyList.add(LIFE_TIME);
+        assertTrue(policyList.contains(LIFE_TIME));
+        assertTrue(policyList.contains(LIFE_TIME_NAME));
+    }
```
Comment (WeiJie96) [inline src/test/java/seedu/address/model/policy/PolicyListTest.java:54]: Perhaps we should test for null and Duplicate

File: src/test/java/seedu/address/model/policy/PolicyListTest.java [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+        assertEquals(policyList, new PolicyList(new Hashtable<>()));
+
+        PolicyList typicalPolicyList = getTypicalPolicyList();
+        assertEquals(typicalPolicyList, new PolicyList(typicalPolicyList.getHashtableCopy()));
+    }
+
+    @Test
+    public void contains() {
+        PolicyList typicalPolicyList = getTypicalPolicyList();
+
```
Comment (WeiJie96) [inline src/test/java/seedu/address/model/policy/PolicyListTest.java:34]: Perhaps we should test for null

File: src/main/java/seedu/address/model/policy/PolicyList.java [hunk 19]
```diff
@@ -22,23 +22,37 @@ public PolicyList() {
...
     public boolean contains(Policy toCheck) {
         requireNonNull(toCheck);
         return policies.contains(toCheck);
     }
 
+    /**
+     * Returns true if the hashtable contains the key as the given argument using PolicyName.
+     */
+    public boolean contains(PolicyName policyName) {
+        return policies.containsKey(policyName.value);
```
Comment (WeiJie96) [inline src/main/java/seedu/address/model/policy/PolicyList.java:36]: Perhaps we should requireNonNull
Same for the other contains()

File: src/main/java/seedu/address/model/policy/PolicyList.java [hunk 20]
```diff
@@ -70,7 +84,7 @@ public static PolicyList createNewPolicyList() {
      * else returns null.
      */
     public Policy getPolicy(PolicyName name) {
-        return policies.get(name.policyName);
+        return policies.get(name.value);
     }
```
Comment (WeiJie96) [inline src/main/java/seedu/address/model/policy/PolicyList.java:89]: Is createNewPolicyList() being used? If it is, we can add it into the PolicyListTest.java

**Review by WeiJie96 — COMMENTED** [review]
Some comments on code quality, testing and other implementation

**Pull request merged**